core-bundle/src/EventListener/ServiceUnavailableListener.php

```diff
@@ -0,0 +1,73 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\CoreBundle\EventListener;
+
+use Contao\CoreBundle\Exception\ServiceUnavailableException;
+use Contao\CoreBundle\Routing\ScopeMatcher;
+use Contao\ManagerBundle\HttpKernel\JwtManager;
+use Contao\PageModel;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpKernel\Event\RequestEvent;
+
+/**
+ * @internal
+ */
+class ServiceUnavailableListener
+{
+    public const JWT_ATTRIBUTE = 'bypass_maintenance';
+
+    private ScopeMatcher $scopeMatcher;
+    private ?JwtManager $jwtManager;
+
+    public function __construct(ScopeMatcher $scopeMatcher, JwtManager $jwtManager = null)
+    {
+        $this->scopeMatcher = $scopeMatcher;
+        $this->jwtManager = $jwtManager;
+    }
+
+    public function __invoke(RequestEvent $event): void
+    {
+        $request = $event->getRequest();
+
+        if (
+            !$this->scopeMatcher->isFrontendMainRequest($event)
+            || $request->attributes->get('_preview', false)
+            || $this->isDisabledByJwt($request)
+        ) {
+            return;
+        }
+
+        $pageModel = $request->attributes->get('pageModel');
+
+        if (!$pageModel instanceof PageModel) {
+            return;
+        }
+
+        $pageModel->loadDetails();
+
+        if ($pageModel->maintenanceMode) {
+            throw new ServiceUnavailableException(sprintf('Domain %s is in maintenance mode', $pageModel->dns));
+        }
+    }
+
+    private function isDisabledByJwt(Request $request): bool
```

Comment:
My problem with this (yes, this also applies to other usages of the `JwtManager` is that if you do not use the `manager-bundle` you cannot use this whole listener. And if @bytehead wanted to provide some custom way to circumvent the maintenance mode, they would have to duplicate the whole listener.
The alternative would be to adjust this listener to check for a certain request attribute and a separate listener with a higher priority in the `manager-bundle` that does set this attribute. Wdyt?

Reply:
Well the same applies to debug mode, does it not? I mean the whole "bypass maintenance through login" will not work anyway because there's no way to set a JWT?

Reply:
Exactly, that's what I'm saying. This should be changed as well. If @bytehead wants to provide this logic, they have absolutely no way of achieving this. Because the `JwtManager` is a concrete class coming from the `manager-bundle` which they cannot use because it does all the ME stuff they don't need.

Reply:
We could also simply dispatch an event 🤔 Then we could use that in the manager-bundle based on the JWT and custom implementation can be easily added. Like...dunno...disable it when logged in using Shibboleth. But then again, we can also implement that once someone needs it.

Reply:
> disable it when logged in 

Right now you won't see the 503 when logged in (backend).

I don't get it, why you check by the JWT token instead if there is a logged in backend user to skip the maintenance mode?

Reply:
The idea is that a client can login on the maintenance page to bypass it. I don't think this needs to be available without a Managed Edition.

Reply:
🤷‍♂️ But login in to backend means still bypass the maintenance page?

Reply:
correct, it is disabled in preview mode (`preview.php`). Not in the regular front end entry point even if logged in to the back end. Not sure if your Symfony installations have a preview endpoint?

Reply:
> Not in the regular front end entry point even if logged in to the back end

But this is the case right now in 4.9 AFAIK.

> Not sure if your Symfony installations have a preview endpoint?

We do have a preview endpoint 👍

Reply:
I noticed this is actually available in any Contao version through the `_bypass_maintenance` flag.